components/service/experiments/src/main/java/mozilla/components/service/experiments/Experiments.kt

```diff
@@ -15,28 +21,91 @@ import mozilla.components.support.base.log.logger.Logger
  * @param valuesProvider provider for the device's values
  */
 @Suppress("TooManyFunctions")
-class Experiments(
-    private val source: ExperimentSource,
-    private val storage: ExperimentStorage,
-    valuesProvider: ValuesProvider = ValuesProvider()
-) {
+open class ExperimentsInternalAPI internal constructor() {
     @Volatile private var experimentsResult: ExperimentsSnapshot = ExperimentsSnapshot(listOf(), null)
     private var experimentsLoaded: Boolean = false
-    private val evaluator = ExperimentEvaluator(valuesProvider)
-    private val logger = Logger(LOG_TAG)
+    private var evaluator: ExperimentEvaluator = ExperimentEvaluator(ValuesProvider())
+    private lateinit var logger: Logger
+
+    private var source: ExperimentSource? = null
+    private var storage: ExperimentStorage? = null
+
+    /**
+     * Initialize the experiments library.
+     *
+     * This should only be initialized once by the application.
+     *
+     * @param applicationContext [Context] to access application features, such
+     * as shared preferences.
+     */
+    fun initialize(
+        applicationContext: Context
+    ) {
+        experimentsResult = ExperimentsSnapshot(listOf(), null)
+        experimentsLoaded = false
+        logger = Logger(LOG_TAG)
+
+        if (source == null) {
```

Comment:
I think it's always helpful to behave sanely (if possible) in the face of calling `initialize` twice.  In that case, we should maybe try to do nothing and log a warning if `source != null`.

Reply:
Currently `initialize()` is called multiple times from tests (and i tried to work around some kinks in testing).
I could use a different pattern though, like a test-only `reset()` function or so?

Reply:
That could work.  That's the pattern used in glean.

Reply:
Yup. Moreover, that would have the nice side-effect of having a developer friendly error log to explain what's going on

Reply:
Ok, so, my challenge is:

**(1)**: `source` & `storage` depend on parameters passed to `initialize()`, so i should set them from there for normal library initialization.

**(2)** I want to enable overriding `source` & `storage` in tests while also allowing repeated `initialize()` calls, e.g.:

```kotlin
test() {
  Experiments.source = FooExperimentSource()
  Experiments.initialize()
}
```

**(3)** I don't want to first set `source` & `storage` from `initialize()`, then override them after, because that will trigger actions later that the test does not expect, i.e.:

```kotlin
test() {
  Experiments.initialize()
  // Oops, starts async operation with default `source` already.
  Experiments.source = FooExperimentSource()
}
```

**4** I want to keep `source` & `storage` internal so i can't expose them as optional parameters to `initialize()`?

```kotlin
test() {
  Experiments.initialize(FooExperimentSource())
  // oops, parameter type can't be internal anymore?
}
```

Reply:
That's what i currently work around with the `null` checks:
```kotlin
test() {
  Experiments.source = FooExperimentSource()
  Experiments.initialize()
  // ok, `source` is non-null, so we won't use the default source.
}

normalProduct() {
  Experiments.initialize()
  // ok, `source` is null, so we will use the default source.
}
```

Reply:
Am i missing a cleaner option here?

Reply:
You could create a test only initialize function with the `@VisibleForTesting` annotation that allows you to inject the source and storage but wouldn't be available to the public API, something like this:
```
@VisibleForTesting(otherwise = VisibleForTesting.NONE)
internal fun initialize(source: ExperimentSource?, storage: ExperimentStorage?) {
     ...
}
```

Reply:
So, another option would be to just use the `ExperimentsInternalAPI` in the tests, so that you can injhect whatever dependency you like for `source` and `storage`. Your code would become something like:

```
class ExperimentsInternalApi(storage, source)
object Experiemnts : ExperimentsInternalApi(new FlatFile(), new KintoSource())
```

In tests you would directly create instances of `ExperimentsInternalApi`

Reply:
Cheers, i will take this to a follow-up bug (as this is a test improvement, separate from reducing the public API surface).